### PR TITLE
WIP add auth routes

### PR DESCRIPTION
**Pull request by velzepooz**

### CODE REVIEW

File: src/auth/auth.service.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+function generateHashForPassword(password) {
+  return new Promise((resolve, reject) => {
+    const salt = crypto.randomBytes(16).toString('base64');
+    crypto.scrypt(password, salt, 64, (err, result) => {
+      if (err) reject(err);
+      resolve(salt + ':' + result.toString('base64'));
+    });
+  });
+}
```
Comment (georgolden) [inline src/auth/auth.service.js]: move the hash function to a separate file. It is crypto utilities

File: src/driver/driver.controller.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * @param {DriverService} driverService
+ * @returns {FastifyRoute[]}
+ */
+export const initDriverController = (driverService) => {
+  const urlPrefix = 'driver';
+
+  const createDriverRoute = {
+    method: 'POST',
+    url: `/${urlPrefix}`,
+    handler: async (req, res) => {
```
Comment (georgolden) [inline src/driver/driver.controller.js]: no validation schema?

File: src/driver/driver.module.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/** @typedef {import('../app.module').AppModule} Module */
+
+import { initDriverService } from './driver.service.js';
+import { initDriverController } from './driver.controller.js';
+
+/**
+ * @return {Module}
+ */
+export const initDriverModule = () => {
+  const driverService = initDriverService({});
```
Comment (georgolden) [inline src/driver/driver.module.js]: use undefined or null is better then empty object if no deps

File: src/driver/dto/driver.dto.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      type: 'string',
+      format: 'date',
+    },
+    updatedAt: {
+      type: 'string',
+      format: 'date',
+    },
+    driverLicenceId: {
+      type: 'string',
+      pattern: '^[A-Z]{2}\\d{7}$',
```
Comment (georgolden) [inline src/driver/dto/driver.dto.js:26]: If use regexp explain in comment what it does

File: test/utils/test.utils.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import fastify from 'fastify';
+import { initAppModule } from '../../src/app.module.js';
+import { config } from '../../src/config.js';
+
+export const buildApp = (opts = {}) => {
```
Comment (georgolden) [inline test/utils/test.utils.js:11]: if there are default options - mention them in the comments at least. If not, why is there an empty object?

**Review by georgolden — CHANGES_REQUESTED** [review]

File: src/driver/driver.controller.js (hunk 2 above)
Comment (velzepooz) [reply src/driver/driver.controller.js]: this is an unused route, I'll just remove it

File: src/auth/jwt.service.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+  try {
+    const { payload } = await jose.jwtVerify(jwtToken, encodedSecret);
+
+    return payload;
+  } catch (e) {
+    console.error(e);
+
+    return null;
+  }
```
Comment (georgolden) [inline src/auth/jwt.service.js]:
```js
return jose.jwtVerify(jwtToken, encodedSecret)
  .then(({ payload }) => payload)
  .catch(console.error);
```

File: src/config.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    database: process.env.POSTGRES_DATABASE_NAME || '',
+    user: process.env.POSTGRES_USER || '',
+    password: process.env.POSTGRES_PASSWORD || '',
+  },
+  jwt: {
+    accessTokenSecret: process.env.ACCESS_TOKEN_SECRET || '',
+    accessTokenExpireTime: process.env.ACCESS_TOKEN_EXPIRE_TIME || '15m',
+    refreshTokenSecret: process.env.REFRESH_TOKEN_SECRET || '',
+    refreshTokenExpireTime: process.env.REFRESH_TOKEN_EXPIRE_TIME || '1d',
+  },
```
Comment (georgolden) [inline src/config.js]: better to split the config by meaning. JWT configuration is for JWT service only, server and db configuration is for infrastructure.

File: src/auth/auth.service.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+ * @typedef {import('../config').Config} Config
+ * @typedef {import('./user.repository').UserRepository} UserRepository
+ * @typedef {import('./jwt.service').JwtService} JwtService
+ * @typedef {object} Deps
+ * @property {UserRepository} userRepository
+ * @property {JwtService} jwtService
+ * @property {Config} config
+ *
+ * @typedef {object} SignInResult
+ * @property {UserDto | null} user
```
Comment (georgolden) [inline src/auth/auth.service.js]:
```js
@property {UserDto} [user]
or
@property {UserDto=} user
```
 * @param {string=} p2 - An optional param (Google Closure syntax)
 * @param {string} [p3] - Another optional param (JSDoc syntax).

File: src/auth/auth.service.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+ * @param {Deps} deps
+ * @param {SignInDto} payload
+ * @return {Promise<SignInResult>}
+ */
+export const signInUser = async (deps, payload) => {
+  const { userRepository, jwtService, config } = deps;
+
+  const user = await userRepository.findOne({ phone: payload.phone });
+
+  if (!user) return { user:  null, accessCookie: '', refreshCookie: '' };
```
Comment (georgolden) [inline src/auth/auth.service.js]: throw Error User not found

File: src/auth/auth.service.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+  const isCorrectPassword = await comparePasswords(payload.password, user.password);
+
+  if (!isCorrectPassword) return { user: null, accessCookie: '', refreshCookie: '' };
```
Comment (georgolden) [inline src/auth/auth.service.js]: throw Error invalid password

File: src/auth/auth.service.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+  return {
+    user: {
+      ...user,
+      password: null,
+    },
+    accessCookie:
+      `Authentication=Bearer ${accessToken}; Secure; HttpOnly; Path=/; Max-Age=${config.jwt.accessTokenExpireTime}`,
+    refreshCookie:
+      `Refresh=${refreshToken}; Secure; HttpOnly; Path=/; Max-Age=${config.jwt.refreshTokenExpireTime}`,
```
Comment (georgolden) [inline src/auth/auth.service.js]: make function for creating Header for token

File: src/auth/auth.service.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+ * @typedef {import('./dto/sign-up.dto').SignUpDto} SignUpDto
+ * @typedef {import('./dto/sign-in.dto').SignInDto} SignInDto
+ * @typedef {import('./dto/user.dto').UserDto} UserDto
+ * @typedef {import('../config').Config} Config
+ * @typedef {import('./user.repository').UserRepository} UserRepository
+ * @typedef {import('./jwt.service').JwtService} JwtService
+ * @typedef {object} Deps
+ * @property {UserRepository} userRepository
+ * @property {JwtService} jwtService
+ * @property {Config} config
```
Comment (georgolden) [inline src/auth/auth.service.js]: you don't need to inject config here. JWT config must be a part of Auth servuce

**Review by georgolden — CHANGES_REQUESTED** [review]
You can use dts files for types. It will reduce amount of typedefs in js files.
E.g.
```js
// @filename: types.d.ts
export type Pet = {
  name: string,
};
 
// @filename: main.js
/**
 * @param { import("./types").Pet } p
 */
function walk(p) {
  console.log(`Walking ${p.name}...`);
}
```

**Comment by georgolden** [comment]
Auth exceptions - User not found and Invalid password must be implemented

File: src/driver/dto/driver.dto.js (hunk 4 above)
Comment (georgolden) [reply src/driver/dto/driver.dto.js:26]:
```js
// driver lisence format. e.g. first 2 letter A-Z, then 7 digits
pattern: '^[A-Z]{2}\\d{7}$',
```
I can give a wrong suggestion please correct

**Review by georgolden — CHANGES_REQUESTED** [review]

File: src/driver/dto/driver.dto.js (hunk 4 above)
Comment (velzepooz) [reply src/driver/dto/driver.dto.js:26]: no, this is correct, I just someway missed you comment during previous review

File: src/auth/auth.exceptions.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  }
+}
+
+export class UnauthorizedException extends AuthException {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+    Error.captureStackTrace(this, this.constructor);
+  }
+}
```
Comment (georgolden) [inline src/auth/auth.exceptions.js]: What is the purpose to have multiple Exception classes?

File: src/auth/auth.controller.js [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+     * @param {FastifyReply} reply
+     * @return {Promise<void>}
+     */
+    handler: async (request, reply) => {
+      const result = await authService.refreshAccessToken(request.userId);
+
+      reply.header('set-cookie', result);
+      reply.code(200);
+    },
+  };
```
Comment (georgolden) [inline src/auth/auth.controller.js]: You can have a hooks folder with your hooks to encapsulate them in a separate file. Please avoid fat controller!

File: src/auth/jwt.guard.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @param {JwtService} jwtService
+ * @return {OnRequestHookHandler}
+ */
+export const initJwtGuard = (jwtService) => async (request) => {
+  const { Authentication } = request.cookies;
+
+  if (Authentication) {
+    /** @type {*} JwtTokenPayload */
+    const data = await jwtService.verifyJwt(Authentication.split(' ')[1], jwtConfig.accessTokenSecret);
```
Comment (georgolden) [inline src/auth/jwt.guard.js]: magic number 1.
```js
// "_" underscore used to ignore a variable
const [_tokenType, token] = Authentication.split(' ');
const data = await jwtService.verifyJwt(token, jwtConfig.accessTokenSecret);
```

File: src/auth/auth.controller.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+import { initJwtGuard } from './jwt.guard.js';
+
+/**
+ * @param {import('../../types/src/auth/auth.service').AuthService} authService
+ * @param {import('../../types/src/auth/jwt.service').JwtService} jwtService
+ * @returns {FastifyRoute[]}
+ */
+export const initAuthController = (authService, jwtService) => {
+  const urlPrefix = '/auth';
+  const jwtGuard = initJwtGuard(jwtService);
```
Comment (georgolden) [inline src/auth/auth.controller.js]: this is overhead, you don't need to init guard or have a guard at all.

File: src/auth/auth.controller.js [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+      const { Refresh } = request.cookies;
+
+      const data = await jwtService.verifyJwt(Refresh, jwtConfig.refreshTokenSecret);
+
+      if (!data) {
+        throw new UnauthorizedException('Unauthorized');
+      }
+
+      request.userId = data.id;
+    },
```
Comment (georgolden) [inline src/auth/auth.controller.js]: You don't use guard here. I don't understand what motivates you to have guard abstraction.

**Review by georgolden — CHANGES_REQUESTED** [review]

**Review by georgolden — APPROVED** [review]
Very nice job! :+1:

**Comment by georgolden** [comment]
Can be merged!

**Pull request merged**